cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -104,6 +115,13 @@ static inline Result<csv::ConvertOptions> GetConvertOptions(
     // Properly set conversion types
     convert_options.column_types[field->name()] = field->type();
   }
+  if (convert_options.include_columns.empty()) {
+    // We know which columns we want - so if none were specified, override the default
+    // behavior of reading all columns by reading a missing fake null column
+    convert_options.include_missing_columns = true;
+    convert_options.include_columns.push_back("fabricated");
```

Comment:
It's probably rare but could this cause an issue if the CSV happened to have a column named `fabricated`?  Although I suppose it'll be addressed by the newline scanning fast implementation.

Reply:
Hmm. I'll have to check, but I think this won't show up even if you do something like `to_table(columns=[])` because then it'll get projected away. Oh, but if you projected a column called `fabricated` that didn't depend on any physical columns…I'll have to rethink this.

Reply:
Ok, now I've added an explicit CSV reader option - it still uses a fabricated null column internally for convenience but this way that won't leak out into consumers.